boot/cmdline.go

```diff
@@ -169,3 +169,76 @@ func ComposeCandidateCommandLine(model *asserts.Model) (string, error) {
 func ComposeCandidateRecoveryCommandLine(model *asserts.Model, system string) (string, error) {
 	return composeCommandLine(model, candidateEdition, ModeRecover, system)
 }
+
+// observeSuccessfulCommandLine observes a successful boot with a command line
+// and takes an action based on the contents of the modeenv. The current kernel
+// command lines in the modeenv can have up to 2 entries when the managed
+// bootloader boot config gets updated.
+func observeSuccessfulCommandLine(model *asserts.Model, m *Modeenv) (*Modeenv, error) {
+	// TODO:UC20 only care about run mode for now
+	if m.Mode != "run" {
+		return m, nil
+	}
+
+	switch len(m.CurrentKernelCommandLines) {
+	case 0:
+		// compatibility scenario, no command lines tracked in modeenv
+		// yet, this can happen when having booted with a newer snapd
+		return observeSuccessfulCommandLineCompatBoot(model, m)
+	case 1:
+		// no command line update
+		return m, nil
+	default:
+		return observeSuccessfulCommandLineUpdate(m)
+	}
+}
+
+// observeSuccessfulCommandLineUpdate observes a successful boot with a command
+// line which is expected to be listed among the current kernel command line
+// entries carried in the modeenv. One of those entries must match the current
+// kernel command line of a running system.
+func observeSuccessfulCommandLineUpdate(m *Modeenv) (*Modeenv, error) {
+	newM, err := m.Copy()
+	if err != nil {
+		return nil, err
+	}
+
+	// get the current command line
+	cmdlineBootedWith, err := osutil.KernelCommandLine()
+	if err != nil {
+		return nil, err
+	}
+	if !strutil.ListContains([]string(m.CurrentKernelCommandLines), cmdlineBootedWith) {
+		return nil, fmt.Errorf("internal error: current command line content %q not matching any expected entry",
```

Comment:
mmm thinking about it, I don't think this is an internal error, so maybe just drop the `internal error:` prefix here maybe